R/dataframe-frame.R

```diff
@@ -2200,3 +2200,114 @@ dataframe__group_by_dynamic <- function(
     )
   })
 }
+
+#' Unstack a long table to a wide form without doing an aggregation
+#'
+#' This can be much faster than a pivot, because it can skip the grouping phase.
+#'
+#' @param ... <[`dynamic-dots`][rlang::dyn-dots]> Column name(s) and selector(s)
+#' to include in the operation. If `NULL` (default), use all columns.
+#' @param step Number of rows in the unstacked frame.
+#' @param how Direction of the unstack. Must be one of `"vertical"` or
+#' `"horizontal"`.
+#' @param fill_values Fill values that don't fit the new size with this value.
+#' This can be a scalar value, an Expr, or a named list of the sort
+#' `list(<column_name> = <fill_value>)`. See examples.
+#'
+#' @inherit as_polars_df return
+#' @examples
+#' df <- pl$DataFrame(x = LETTERS[1:8], y = 1:8)$with_columns(
+#'   z = pl$int_ranges(pl$col("y"), pl$col("y") + 2, dtype = pl$UInt8)
+#' )
+#' df
+#'
+#' df$unstack(step = 4, how = "vertical")
+#' df$unstack(step = 2, how = "horizontal")
+#' df$unstack(cs$numeric(), step = 5, fill_values = 0)
+#' df$unstack("x", "y", step = 5, fill_values = list(y = 999, x = "foo"))
+dataframe__unstack <- function(
+  ...,
+  step,
+  how = c("vertical", "horizontal"),
+  fill_values = NULL
+) {
+  wrap({
+    check_dots_unnamed()
+
+    is_named_list <- is.list(fill_values) &&
+      !is.null(names(fill_values)) &&
+      any(names(fill_values) != "")
+    is_scalar_or_expr <- length(fill_values) == 1 || is_polars_expr(fill_values)
+    if (!is.null(fill_values) && !is_named_list && !is_scalar_or_expr) {
+      abort("`fill_value` must be a scalar or a named list.")
+    }
+
+    if (!(is_scalar_integerish(step) && isTRUE(step > 0L))) {
+      abort("`step` must be a single positive integer-ish value")
+    }
+    how <- arg_match0(how, values = c("vertical", "horizontal"))
+
+    dots <- list2(...)
+    df <- if (length(dots) == 0L) {
+      self
+    } else {
+      self$select(!!!dots)
+    }
+
+    height <- self$height
+    if (how == "vertical") {
+      n_rows <- step
+      n_cols <- ceiling(height / n_rows)
+    } else {
+      n_cols <- step
+      n_rows <- ceiling(height / n_cols)
+    }
+
+    # We know whether unstacking will create empty cells, so we extend the
+    # data (and potentially fill missing values) before unstacking.
+    n_fill <- n_cols * n_rows - height
+    if (n_fill > 0) {
+      if (is.null(fill_values)) {
+        fill_values <- NA
+      }
+      if (!is.list(fill_values)) {
+        fill_values <- rep(list(fill_values), df$width)
+        names(fill_values) <- names(df)
```

Comment:
This is to allow passing columns in `fill_values` in a different order than in `...`, e.g. `unstack("a", "b", step = 2, fill_values = list(b = 1, a = 2))`